packages/next/src/client/components/use-reducer-with-devtools.ts

```diff
@@ -149,16 +124,28 @@ function useReducerWithReduxDevtoolsImpl(
     )
     if (devtoolsConnectionRef.current) {
       devtoolsConnectionRef.current.init(normalizeRouterState(initialState))
+
+      if (actionQueue) {
+        actionQueue.devToolsInstance = devtoolsConnectionRef.current
+      }
     }
 
     return () => {
       devtoolsConnectionRef.current = undefined
     }
-  }, [initialState])
+  }, [initialState, actionQueue])
+
+  const dispatch = useCallback(
+    (action: ReducerActions) => {
+      if (!actionQueue.state) {
+        // we lazy initialize the mutable action queue state since the data needed
+        // to generate the state is not available when the actionQueue context is created
+        actionQueue.state = initialState
```

Comment:
This seems a bit fishy to me, see https://github.com/vercel/next.js/pull/56497/files#r1362317893

It seems like we should be able to hoist the state initialization outside of the render phase entirely